Custom dyld linker for iOS mach-o executable files

apeskov opened the pull request:
### Description

This PR is a part of iOS enabling efforts. Particularly it allows to use regular TVM RPC upload mechanics in the way like for Android platform.

Essential part of auto tuning on real iOS device is an ability to run kernel from module received over network. This patch allows to load shared libraries received by RPC server from host side and stored in temporal directory on device. No requirements of signing this binary.

### Details of patch
- New one `dso_loader` plugin specially for iOS RPC app.
- The `dso_loader` itself is built on top of `macho_dyld` external module (under Apple Public Source License Version 2.0).
- Dependency is downloaded automatically via CMake external project mechanics.
- Added `USE_IOS_RPC` flag to enable build iOS `tvmrpc` app inside main TVM crake project 
- Fixed compilation and warnings(suppressed) for `tvmrpc` Xcode project
- Add one more test to check `custom_dso_loader` plugin. Previous tests is still working
- Add one more env var `TVM_IOS_TEAM_ID`. Allow to avoid requirements to call `init_proj.py` each time before test run.

### Details of dyld module
Modified and simplified version of original dyld library.
* Provide interfaces similar to dlopen, dlclose, dlsym and dlerror functions. Function names is prefixed with "custom_".
* API header is under ALv2. Sources is under Apple Public Source License.
* Check of binary signing is switched off.
* Works only under debugger (lldb). The JIT itself is not allowed for iOS, but it has some mitigation for debuggers.
* Provides very limited functionality. A lot of features are disabled or limited.

tqchen commented:
Thanks @apeskov , it would be great to know if we can rely on the apple DYLD via submodule or linked library. Additionally, we want to check if the license(Apple license) is compatible with ALv2. 

If not, we cannot checkin the code that directly relies on this  and would need to create a separate repo for the iOS RPC .

tqchen commented:
After looking at the source code, I think it would be better to isolate the piece of macho-dyld linker as a separate repo while exposing a set of common C APIs that we can depend on, in particular the following C APIs.

- macho_dlopen
- macho_dlsym
- macho_dlclose

We can then depend on these API headers cleanly(as they are common public API that can be reimplemented in other platforms). This would alleviate the possible license mixing concerns that might arise.

apeskov commented:
Reopen. Was closed by mistake.

apeskov commented:
@tqchen Your recommendation was fulfilled.
* The conflict sources was extracted into separate repo https://github.com/octoml/macho-dyld
* TVM is linked with macho_dyld dynamically
* TVM depends only on one header file `macho_dlfcn.h`

Some details about macho-dyld. All sources inside this repo including written by myself have an Apple license header (I treat it as deep modification of original implementations). There is only one header `macho_dlfcn.h` without any license it has an declaration of API functions. I hope that is a correct handling of Apple sources in context of integration to TVM.

tqchen commented:
Thanks @apeskov . This is an important step. Would be great if we can think about de-couple further to reverse the dependency. (e.g. make the macho depend on libtvm as a plugin).

The main rationale is that depending on the license we may not want to directly make this as a submodule(which means we will package and distribute the source by default and there might be some license complications).

Here is an idea:
- We create another plugin module that depends on libtvm which registers an overrided version of so loader.
- see https://github.com/apache/tvm/blob/main/src/runtime/dso_library.cc#L93
- We can then run a plugin that registers the new loader back to TVM. see an example here https://github.com/apache/tvm/tree/main/apps/extension

This way we do not need to hijack the dso library

apeskov commented:
Thank you @tqchen for comment! Your concerns about license mixing is relevant. And I absolutely agree that keep license clean is a priority.

I've got your point about reverse linker->tvm dependency. And moreover It will be more elegant solution I guess. I will do it on iOS RPC server side via "tvm.rpc.server.load_module" rewriting. It will have the same effect. 

But anyway, I didn't fully catch your position regarding keep this module in 3rdparty folder. Is it acceptable or not? My wishes is very simple, I would like to have TVM tuning with iOS devices out of box. Doesn't matter how it will be distributed: integrated submodule, checked out by cmake or installed from brew.  

If we would like to keep 3rdparty folder only for Apache/MIT/BSD licensed code, I can suggest next plan of integration:

1. Move all changes into iOS RPC server app. New DSO loader injection via "tvm.rpc.server.load_module" modification.
2. Implement minimal loader version from scratch and put it under ALv2 license near iOS_rpc app. Just a minimal quasi loader  suitable for tuning purposes. It will depend only on system header files from public SDK (like `mach-o/loader.h` and may be some others). I hope it's applicable to include files with Apple License header from apache code.

@tqchen What do you think regarding this?

tqchen commented:
Thanks @apeskov I agree moving the changes into the iOS rpc app.  The files we include in the 3rdparty would beed to be apache compatible, because we bundle and release the source code, which requires more careful handling.

Only putting macho loader that is ALv2, and provide instructions/script for the user of iOS RPC to download and link the macho loader is cleaner. Since from the source release point of view we are not bundling the source code

apeskov commented:
@tqchen I've tried to address your comments. Could you please take a look on updated version of this PR and provide a feedback is it OK or not.

As in short:
* I've moved macho-dyld repository into cmake level as ExternalProject. It will be automatically downloaded and built with `ios_rpc` cmake target. This targets are enabled under special flag "USE_IOS_RPC" (is switched off by default).
* Xcode project is modified to support conditional compilation of custom_dso_loader plugin triggered by user defined setting "USE_CUSTOM_DSO_LOADER=1". Otherwise it will work as before.
* Corresponding section in readme and test scripts are also added.

apeskov commented:
@tqchen @jcf94 Could you please continue review or assign someone to take a look.

tqchen commented:
Thanks @apeskov !
